### PR TITLE
Pass mention list down to puppet and use mentionIdList from puppet if possible

### PR DESCRIPTION
**Pull request by windmemory**
Refer to issue: https://github.com/Chatie/wechaty/issues/1560

### CODE REVIEW

File: src/user/message.ts [hunk 1]
```diff
@@ -515,7 +517,8 @@ export class Message extends Accessory implements Sayable {
       await this.puppet.messageSendText({
         contactId: mentionContact.id,
         roomId: room.id,
-      }, textMentionList + ' ' + text)
+      }, textMentionList + ' ' + text,
```
[the remark this answers is not in the document]
Comment (windmemory) [reply src/user/message.ts]: Okay, will do

File: src/user/message.ts [hunk 2]
```diff
@@ -51,6 +51,8 @@ import {
   UrlLink,
 }                       from './url-link'
 
+import { MessagePayloadBase, MessagePayloadRoom } from 'wechaty-puppet/dist/src/schemas/message'
```
[the remark this answers is not in the document]
Comment (huan) [reply src/user/message.ts]: I understand the problem but I have no time to investigate in it right now.

Please feel free to fix it in your way, as long as you believe it's a best practice solution.

[the remark this answers is not in the document]
Comment (windmemory) [reply src/user/message.ts]: Okay, will submit change, thanks for squeezing time for my PR :)
U R the best

**Comment by windmemory** [comment]
Updated

**Comment by huan** [comment]
Please turn the CI green before we merge this PR, thanks.

**Comment by windmemory** [comment]
I've fixed 4 ci jobs, the 2 remain failed. 

#### Travis CI

The Travic CI build failed in the process of running `npm run test:pack`, and I saw the error is below:
```shell
02:14:56 INFO PuppetManager install(wechaty-puppet-puppeteer@^0.14.1) done
02:15:27 ERR PuppetPuppeteerBridge start() exception: TimeoutError: Navigation Timeout Exceeded: 30000ms exceeded
02:15:27 ERR PuppetPuppeteer initBridge() exception: TimeoutError: Navigation Timeout Exceeded: 30000ms exceeded
02:15:27 WARN PuppetPuppeteerBridge stop() page.close() exception: Error: Protocol error: Connection closed. Most likely the page has been closed.
02:15:27 ERR PuppetPuppeteer start() exception: Error: Error: TimeoutError: Navigation Timeout Exceeded: 30000ms exceeded
Error: Error: Error: TimeoutError: Navigation Timeout Exceeded: 30000ms exceeded
    at PuppetPuppeteer.puppet.on.error (/tmp/npm-pack-testing.6486/node_modules/wechaty/dist/src/wechaty.js:437:44)
    at PuppetPuppeteer.emit (events.js:182:13)
    at PuppetPuppeteer.EventEmitter.emit (domain.js:442:20)
    at PuppetPuppeteer.emit (/tmp/npm-pack-testing.6486/node_modules/wechaty-puppet/dist/src/puppet.js:165:22)
    at PuppetPuppeteer.<anonymous> (/tmp/npm-pack-testing.6486/node_modules/wechaty/node_modules/wechaty-puppet-puppeteer/dist/src/puppet-puppeteer.js:106:22)
    at Generator.throw (<anonymous>)
    at rejected (/tmp/npm-pack-testing.6486/node_modules/wechaty/node_modules/wechaty-puppet-puppeteer/dist/src/puppet-puppeteer.js:23:65)
02:15:27 ERR Wechaty start() exception: Error: Error: TimeoutError: Navigation Timeout Exceeded: 30000ms exceeded
Error: Error: Error: TimeoutError: Navigation Timeout Exceeded: 30000ms exceeded
    at PuppetPuppeteer.puppet.on.error (/tmp/npm-pack-testing.6486/node_modules/wechaty/dist/src/wechaty.js:437:44)
    at PuppetPuppeteer.emit (events.js:182:13)
    at PuppetPuppeteer.EventEmitter.emit (domain.js:442:20)
    at PuppetPuppeteer.emit (/tmp/npm-pack-testing.6486/node_modules/wechaty-puppet/dist/src/puppet.js:165:22)
    at PuppetPuppeteer.<anonymous> (/tmp/npm-pack-testing.6486/node_modules/wechaty/node_modules/wechaty-puppet-puppeteer/dist/src/puppet-puppeteer.js:106:22)
    at Generator.throw (<anonymous>)
    at rejected (/tmp/npm-pack-testing.6486/node_modules/wechaty/node_modules/wechaty-puppet-puppeteer/dist/src/puppet-puppeteer.js:23:65)
02:15:27 INFO Wechaty <wechaty-puppet-mock> stop() v0.23.22 is stoping ...
02:15:27 INFO Wechaty <wechaty-puppet-wechat4u> stop() v0.23.22 is stoping ...
02:15:27 INFO Wechaty <wechaty-puppet-puppeteer> stop() v0.23.22 is stoping ...
02:15:27 WARN PuppetPuppeteerBridge stop() page.close() exception: Error: Protocol error: Connection closed. Most likely the page has been closed.
```

Since the error is related to `wechaty-puppet-puppeteer`, I got no clue of how to fix it. Could you please help me on that?

#### code climate

It complains about:
```
Function mention has a Cognitive Complexity of 6 (exceeds 5 allowed). Consider refactoring. 
```
Seems like this issue needs a refactor work on the `mention` function, which might introduce a lot of changes potentially. And I saw some other improvements could be done for the `message.ts` file in the `codeclimate` [console](https://codeclimate.com/github/Chatie/wechaty/src/user/message.ts/source?from_sha=5abfc704&to_sha=84026611). So I would suggest that we create another issue to track this work instead of in this one.

**Comment by huan** [comment]
Ok, please file another issue to track CodeClimate.

And could you please push another commit to see if the Travis CI could pass?

**Comment by windmemory** [comment]
> Ok, please file another issue to track CodeClimate.
Sure, will create another issue for the CodeClimate issue.

> And could you please push another commit to see if the Travis CI could pass? Because the puppeteer puppet sometimes will timeout for no reason.
No, Travis CI won't pass, I tried to restart the Travis CI job manually, but failed again with same error.

**Comment by huan** [comment]
Do not restart Travis ci manually. Push another commit should be better.

**Comment by windmemory** [comment]
Got it, will do.

**Comment by windmemory** [comment]
While reviewing another PR, I noticed that I forgot to modify the code in `room.ts` file... Will do that in a minute.

**Comment by huan** [comment]
Yes, there's something caused the confusion.

After having a look into the code, I believe there should not be many places to support the `mentionId`. One and only one method support it should be the best practice.

Then I'd like to do the following change:

```diff
class Message {
  // ...
  public async say (	   public async say (
    textOrContactOrFileOrUrl : string | Contact | FileBox | UrlLink,	     textOrContactOrFileOrUrl : string | Contact | FileBox | UrlLink,
-     mention?   : Contact | Contact[],	     mention?   : Contact | Contact[],
  ): Promise<void> {
```

Remote the `metion` support in the `Message.say`, only leave it in the `Room.say`, because it's room related.

**Comment by windmemory** [comment]
Actually I don't quite agree with the design.

When I use `message.say()` I might expect that I am replying this message. So if this happened in a room, I would like to mention the one who said the message. Does this make sense to you?

If so, we should have a default mention contact in the `message.say()` when it happened in a room.

**Comment by huan** [comment]
It's wired: why do I have to mention someone when I say something in a room?

**Comment by windmemory** [comment]
If you call the `message.say()` function, then you already got a message, and this message is from a room. When you reply back this message, you want to reply to the people who said this message.(I assume)

Imagine that multiple people discuss something in a room, and only someone attracts your attention, and you only want to reply to that message.

**Comment by windmemory** [comment]
I will assume that if you want to say something in a room, you will use `room.say()` not `message.say()`.

**Comment by huan** [comment]
No, _reply_ is not the case.

`say` is just you want to say something. it's named as `say` not a `reply` because it's not a reply function.

the `Message.say` is just a convenience method to let you not have to specify the room or contact conversation.

Just say it is enough.

### OR

How about let's get rid of the `Message.say` Entirely? Because it seems to make you a bit confusing, maybe others will too.

**Comment by windmemory** [comment]
Yep, if that's the definition of `message.say()`, then we should not add extra logic in it.

We can remove it, or rename it as `reply`? `Message.reply()` makes more sense, and give developer convenience to do the reply logic in it.

**Comment by windmemory** [comment]
Removed `mention` in the `Message.say` and pass the 5 CIs.

**Comment by huan** [comment]
Thanks for the great contribution!

**Comment by windmemory** [comment]
Glad to see `Wechaty` getting better :)

**Comment by windmemory** [comment]
Just a small reminder, `Label` related design is still in pending, I am waiting to implement that feature :P

Besides, @xinbenlv is eager to have that feature in `Wechaty`, so please squeeze a little time review that design~

**Pull request merged**